Kubernetes/HOWTO-on-prem.md

```diff
@@ -0,0 +1,171 @@
+# Kubernetes On-Premises Deployment From Scratch -- Start to Finish #
+This guide will walk you through deploying *Kubernetes 1.8* on a Linux master and join two Windows nodes to it without a cloud provider.
+
+## Assumptions and Prerequisites ##
+
+It is assumed you will be setting up the following:
+
+- 1 Linux VM/host on an Ubuntu like OS. This will be the single master
+- 2 Windows Server 2016 RS3+ VM/host as worker nodes
+- Compilation of required binaries specific to this exercise can be done on the Linux host above, or a temporary Linux host via Vagrant, custom VM, or otherwise.
+- Cluster subnet is assumed to be 192.168.0.0/16 but could be specified otherwise. Changes to this value are not covered.
+- Service subnet is assumed to be 11.0.0.0/16 and is hardcoded throughout this repo. Changes to this value are not covered.
+
+A few prerequisite definitions and requirements for networking:
+
+- The **external network** is the physical network across which nodes communicate. This exists regardless of whether or not you follow this guide.
+- The **cluster subnet** is a (<a href="#allow-routing">routable</a>) virtual network that must be a /16. Each _node_ will grab a /24 from the subnet to use for its pods.
+- The **service subnet** is a hardcoded 11.0/16 network that is translated into cluster space by `kube-proxy` running on the node.
+
+**Note**: The guide will assume a local working directory of `~/kube` for the Kubernetes setup. If you choose a different directory, just replace any references to that path.
+
+## Building the Binaries ##
+We will begin with a Linux master node on a recent Ubuntu-like distro and a Windows worker node on RS3+; all of these are assumed to be VMs.
+
+First, install all of the pre-requisites on the master:
+
+    $ sudo apt-get install git build-essential docker.io conntrack
+
+### Building Kubernetes ###
+We will need to build the `kubelet` and `kubeproxy` binaries for Windows from scratch _cross-compiling from Linux_, so is a non-issue by leveraging the [standard containerized build scripts](https://github.com/kubernetes/kubernetes/tree/master/build#key-scripts) leveraged in the Kubernetes project.
+
+Various permission denied errors are avoided by building the Linux kubelet first, per the note at [acs-engine](https://github.com/Azure/acs-engine/blob/master/scripts/build-windows-k8s.sh#L176):
+
+> Due to what appears to be a bug in the Kubernetes Windows build system, one has to first build a Linux binary to generate `_output/bin/deepcopy-gen`. Building to Windows w/o doing this will generate an empty `deepcopy-gen`.
+
+You can optionally execute this in a Vagrant box:
+
+```bash
+DIST_DIR="${HOME}/kube/"
+mkdir ${DIST_DIR}
+SRC_DIR="${HOME}/src/kubernetes/"
+mkdir -p "${SRC_DIR}"
+git clone https://github.com/madhanrm/kubernetes.git ${SRC_DIR}
+cd ${SRC_DIR}
+git checkout cniwindows
+build/run.sh make WHAT=cmd/kubelet KUBE_BUILD_PLATFORMS=linux/amd64

```

Comment:
Following these instructions (from scratch and exactly) I had an issue with this build/run.sh command because docker was installed by apt but the service wasn't started. Then after starting the service with 'sudo service docker start' the command still failed because I was not root or using sudo + my user was not in the 'docker' group.

Reply:
I haven't actually tried the dockerized steps, so thanks. I'll link to [this answer](https://askubuntu.com/a/477554).